src/main/java/com/example/gradecalculator/service/SubjectService.java

```diff
@@ -0,0 +1,149 @@
+        this.userSubjectRepository = userSubjectRepository;
+        this.subjectRepository = subjectRepository;
+        this.userRepository = userRepository;
+        this.selectedSubjectsByYear = new HashMap<>();
+        this.schoolYearRepository = schoolYearRepository;
+    }
+
+    public SubjectTO dataTO(Subject subject) {
+        if (subject == null) {
+            return null;
+        }
+        return new SubjectTO(subject.getId(), subject.getName(), subject.getDescription());
+    }
+
+    public List<SubjectTO> getAllSubjects() {
+        return StreamSupport.stream(subjectRepository.findAll().spliterator(), false)
+                .map(this::dataTO)
+                .sorted()
+                .collect(Collectors.toList());
+    }
+
+    public Set<UserSubject> getUserSubjectsByYearAndUserId(int year, Long userId) {
+        return userSubjectRepository.findBySchoolYearAndUserId(year, userId);
+    }
+
+    public void selectSubjectForYear(int year, String subject) {
+        Set<String> selectedSubjects = selectedSubjectsByYear.getOrDefault(year, new HashSet<>());
+        selectedSubjects.add(subject);
+        selectedSubjectsByYear.put(year, selectedSubjects);
+    }
+
+    public Set<String> getSelectedSubjectsForYear(int year) {
+        return selectedSubjectsByYear.getOrDefault(year, new HashSet<>());
+    }
+
+    public void removeSubjectForYear(int year, String subject) {
+        Set<String> selectedSubjects = selectedSubjectsByYear.get(year);
+        if (selectedSubjects != null) {
+            selectedSubjects.remove(subject);
+            if (selectedSubjects.isEmpty()) {
+                selectedSubjectsByYear.remove(year);
+            }
+        }
+    }
+
+    public List<String> saveSubjects(String[] selectedValues, long userId) {
+
+        var errors = new ArrayList<String>();
+        for (String entry : selectedValues) {
+            var splittedString = entry.split("-");
+            long selectedSubjectId = Long.parseLong(splittedString[0]);
+            long selectedYearId = Long.parseLong(splittedString[1]);
+            String error = saveSubjectYearSelectedUser(userId ,selectedSubjectId, selectedYearId);
+            if(error != null){
+                errors.add(error);
+            }
+        }
+
+        return errors;
+    }
+
+    public String saveSubjectYearSelectedUser(long userId, long subjectId, long yearId) {
+
+        Subject selectedSubject = subjectRepository.findById(subjectId).orElseThrow(() -> new IllegalArgumentException("Subject not found"));
+        SchoolYear selectedYear = schoolYearRepository.findById(yearId).orElseThrow(() -> new IllegalArgumentException("Year not found"));
+        User selectedUser = userRepository.findById(userId).orElseThrow(() -> new IllegalArgumentException("User not found"));
```

Comment:
this is quite inefficent as the method is currently only used in a loop for the same user.
-> user should only be loaded once and given to this method as parameter
Also you could check for duplicated subject or years, so you only load every entry exactly once and not possible multiple times